---
title: "wnv_omics_pretty"
author: "Aeriel Belk, Amanda Walz, Rebecca Cheek, & Kyle Root"
date: "December 5, 2018"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, error = FALSE, comment = FALSE)
options(readr.show_progress = FALSE)
```

```{r libraries}
library(tidyverse)
library(ggthemes)
library(seqinr)
library(purrr)
library(dplyr)
library(readr)
library(phyloseq)
library(stringr)
library(tidyr)
library(lubridate)
library(forcats)
library(tigris)
library(knitr)
library(ggplot2)
library(sf)
library(ggdendro)
library(leaflet)
library(viridisLite)
library(RColorBrewer)
library(dendextend)

# if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
# BiocManager::install("Biostrings", versilibrary(devtools) on = "3.8"

library(Biostrings)

# if (!requireNamespace("BiocManager", quietly=TRUE))
# install.packages("BiocManager")
# BiocManager::install("msa")

library(msa)

# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("DECIPHER", version = "3.8")
library(DECIPHER)


```

# Introduction


## Sample Collection
Samples were collected from both bird tissues __(n = 222)__ and pooled mosquito populations __(n = 419)__.   

```{r county_metadata}
wn_meta_filtered <- read.csv("../metadata/metadata.csv") %>% 
  select(Scripps.ID, Species, Sample.type, Collection.date, County, Latitude, Longitude)

wn_meta_sum_county <- wn_meta_filtered %>% 
  group_by(County) %>% 
  summarize(
    n = n()) %>% 
  ungroup()

wn_meta_sum_county %>%
  mutate(County = fct_reorder(County, n)) %>% 
  ggplot(aes(x = County, y = n)) +
  geom_point() +
  coord_flip() +
  theme_minimal() +
  labs(title = "Samples collected by county", x = NULL, y = "# of samples")
```


```{r bird_metadata}
wn_meta_sum_species <- wn_meta_filtered %>% 
  filter(Sample.type == "Tissue") %>% 
  group_by(Species) %>% 
  summarize(
    n = n()) %>% 
  ungroup()

wn_meta_sum_species %>%
  mutate(Species = fct_reorder(Species, n)) %>% 
  ggplot(aes(x = Species, y = n)) +
  geom_point() +
  coord_flip() +
  theme_minimal() +
  labs(title = "Samples collected by bird species", x = NULL, y = "# of samples")    
```


```{r mosquito_metadata}
wn_meta_sum_mosquitos <- wn_meta_filtered %>% 
  filter(Sample.type == "Pooled mosquitos") %>% 
  group_by(Species) %>% 
  summarize(
    n = n()) %>% 
  ungroup()

wn_meta_sum_mosquitos %>%
  mutate(Species = fct_reorder(Species, n)) %>% 
  ggplot(aes(x = Species, y = n)) +
  geom_point() +
  coord_flip() +
  theme_minimal() +
  labs(title = "Samples collected by mosquito species", x = NULL, y = "# of samples")    
```

********************************************************************************

# Genomics
## Rebecca's part





## Comparative Dendrogram
  
  Consensus sequences were obtained from the Andersen lab github. These sequences, collected from a variety of sources including several bird and mosquito species, were aligned against each other so we could determine point mutation differences in the sequences that could be attributed to distinct strains of West Nile Virus. The  hypothesize investigated here is that sequences isolated from different organisms may be phylogenetically different. This is likely based on general knowledge of virus evolution. In general, evolutionary changes occur in response to pressures from the host immune system or other environmental effects. Thus, it is likely that strains that begin to differentiate in a bird may be different than those that begin to differentiate in mosquitos.
  To investigate this, the individual species of birds were categorized into four larger groups: corvids, hawks, songbirds, and mosquitos. Then, the aligned strains were hierarchically clustered based on Hamming distance using the R package biostrings. These clusters were visualized as a colored dendrogram to evaluate whether the sequences from individual species groups were more similar to each other than others.
  
```{r dendrogram}

# import dataset that Rebecca made:
clean_seq <- readDNAStringSet("../fasta/clear_wnv_clean.fas")

#import metadata
metadataframe <- read_csv('../metadata/metadata.csv')
metadataframe <- metadataframe %>% 
  rename_all(tolower)
colors <- metadataframe$color

# unmasked clustering
sdist <- stringDist(as(clean_seq,"DNAStringSet"), method="hamming")
clust <- hclust(sdist, method = "ward.D2")
dend <- as.dendrogram(clust)
dend_data <- dendro_data(dend, type = 'rectangle')
  
dend <- assign_values_to_leaves_edgePar(dend, value = colors, edgePar = "col") # add colors
plot(dend, leaflab = 'none', horiz = TRUE, axes = FALSE)

```
  
  Colors: 
  
  - Blue = corvid
  - Green = hawk
  - Red = mosquito
  - Pink = songbird
  
  Based on this visualization, it is clear that the sequences did cluster by origin species. There were a greater number of corvid and mosquito samples, so these are predominant on the plot, making it difficult to interpret differences in other species. Interestingly, the first break separated a group of corvids from the other species, though not all of the corvid strains were different at this break. The lower levels of the plot further separate the strains. The mosquitos all cluster together in the center of the plot, which was expected. The small group of hawk species also clusterered together, separated at lower levels from the otehrs. The songbird group is also clustered seperately, but appear to be more closely related to corvids than mosquitos, which is interesting. A deeper understanding of the relatedness of West Nile Virus between species might illuminate transmission patterns in the disease that could help researchers contain the illness and prevent futher spread.
  
*******************************************************************************
  
  


