---
title: "â€˜Omics Project"
author: "Aeriel Belk, Amanda Walz, Rebecca Cheek, Kyle Root"
date: "11/7/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This is an R-markdown we can use!!

Steps: 

1. import and early explore data
2. then, see what analyses we can do

```{r}
library(readr)
wn_meta<- read.csv("metadata/metadata.csv")
```
 

Ideas for us to look at. 
Ariel: Build a phylogenetic tree using Bioconductor(phyloseq) to look at divergence and host species
Kyle:Compare genetic viersity between host species and vecotr species
Amanda: 
Rebecca: Landscape genetics approach to look at how the landscape affects genetic structure or rate of gene flow. 


Do some analysis of the metadata:

 - species
 - latitude/longitude
 - distribution over counties

``` {r metadata}



```


genomic analysis

 - phyloseq?
 
```{r genomics}
library(phyloseq)


```
