---
title: "â€˜Omics Project"
author: "Aeriel Belk, Amanda Walz, Rebecca Cheek, Kyle Root"
date: "11/27/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This is an R-markdown we can use!!

Steps: 

1. import and early explore data
2. then, see what analyses we can do

```{r}
library(readr)
library(stringr)
wn_meta<- read.csv("metadata/metadata.csv")
```
 

Ideas for us to look at. 
Aeriel: Build a phylogenetic tree using Bioconductor(phyloseq) to look at divergence and host species
Kyle:Compare genetic viersity between host species and vecotr species
Amanda: sf plot to use lat-long data and group smilar sequences by region
Rebecca: Landscape genetics approach to look at how the urban vs. rural affects genetic structure or rate of gene flow (see https://doi.org/10.1111/mec.14783). Maybe use popualtion density in each county as a proxy for urbanization.  


Do some analysis of the metadata:

 - species 
 - latitude/longitude
 - distribution over counties
 - sequence similarities/ density by region
 
```{r libraries, message=FALSE, warning=FALSE}
library(seqinr)
library(purrr)
library(dplyr)
library(readr)
library(phyloseq)
library(stringr)
library(tidyr)
library(lubridate)
library(forcats)
library(tigris)
library(knitr)
library(ggplot2)
library(sf)

# if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
# BiocManager::install("Biostrings", versilibrary(devtools) on = "3.8"

library(Biostrings)

if (!requireNamespace("BiocManager", quietly=TRUE))
install.packages("BiocManager")
BiocManager::install("msa")

library(msa)

```

``` {r metadata}
metadata_url <- paste0("https://raw.githubusercontent.com/aeriel-belk/NSCI535-westnile-omics/master/metadata/metadata.csv")

metadataframe <- read_csv(metadata_url)
metadataframe <- metadataframe%>% 
  rename_all(tolower)

```


steps to read in:
1) list.files with full names option to get names of all files in folder, save in object (char vector)
2) pipe that vector into map function that does read.fasta (play around with just the first file first)
3) result should be a long list

seqinr -> read.fasta

```{r read-in}
#1) list files into vector
files <- list.files(path = 'consensus_sequences', full.names = TRUE)

#2) make map function to read in fasta files
#fasta <- map(.x = files, .f = read.fasta, 
#             as.string = TRUE, forceDNAtolower = FALSE)

fasta2 <- readDNAStringSet(files, format="fasta", nrec=-1L, skip=0L, seek.first.rec=FALSE, use.names=TRUE)

writeXStringSet(fasta2, 'consensus-seqs.fna')
#writeXStringSet(LargeDNAStringSet, 'fname.fa')
#write.fasta(sequences, names, file.out, open = "w", nbchar = 60, as.string = FALSE)

```

*******************************************************************************

### Align Sequences
  
Using BioStrings (https://bioconductor.org/packages/release/bioc/html/Biostrings.html) to align sequences so we can compare across samples. Plan is to align, then try to cluster the sequences into groups based on similarity. This will allow us to build trees, which we can try to apply to Amanda and Rebecca's analyses of season and population, as well as host species.
  
  After getting it important and *roughly* in a usable format, I tried doing some alignment and clustering with the biostrings package
  
```{r biostring, eval=FALSE}
#sdist <- stringDist(as(fasta2,"DNAStringSet"), method="hamming")
#clust <- hclust(fasta2, method = "single")

#sdist <- stringDist()

### 11-28-18, ADB ###

aligned <- DNAMultipleAlignment(fasta2)
# unequal number of characters error

aligned2 <- readDNAMultipleAlignment('consensus_sequences', 'fasta')
# got output with 0 rows, but a start!





```
  
  
********************************************************************************
  
*Amanda tenative plan to make an interactive map characterizing WNV infections by season, species*

```{r mapped_plot}

#Tidy data for county map with infections by season
ca_wnv <- metadataframe %>% 
  select("species", "collection date",
           "county", "latitude", "longitude") %>% 
  rename_all(funs(str_replace(., "\\s", "_"))) %>% 
  separate(col = collection_date, 
           into = c("month", "day", "year"),
           sep = "/") %>% 
  mutate(month = factor(month),
         season = fct_collapse(month,
               winter = c("12", "1", "2"),
               spring = c("3", "4", "5"),
               summer = c("6", "7", "8"),
               fall = c("9", "10", "11"))) %>% 
  mutate(latitude = as.double(latitude)) %>% 
  mutate(longitude = as.double(longitude)) %>% 
  group_by(latitude) %>% 
  na.omit()

ca_wnv_sf <- st_as_sf(ca_wnv, 
                      coords = c("latitude", "longitude"),
                      crs = 3310)

season_summary <- ca_wnv %>% 
  group_by(season) %>% 
  summarize(no_infections = n())
kable(season_summary)

#Creating CA sf boundaries
ca_counties <- counties(state = "CA", cb = TRUE, class = "sf")
ggplot() +
  geom_sf(data = ca_counties, aes(fill = AWATER)) +
  scale_fill_viridis_c(name = "Water area") +
  facet_wrap(~ season, ncol = 4)


```
  
********************************************************************************
*Rebecca making a map to look at urbaization and WNV genetic structure*

Trying to relate the virus's genetic structure to urbanization. Since the data doesn't provide a real gradient of 'urbanization' we will use human popualtion density as a proxy
```{r}
####Human population density map bit##
#### Rebecca trying to make life easier by using census data per county as a proxy for urbanization
library(tidyverse)
library(tidycensus) ##tidycensus package
library(leaflet)
library(stringr)
library(sf)
library(adegenet)
library(pegas)
library(Biostrings)
library(dplyr)
library(seqinr)
library(ape)
library(ShortRead)

census_api_key("a22c251a387f5312b9e5c2e5ad21494088e984f6")

ca_pop <- get_acs(geography = "county", 
                     variables = "B01003_001", 
                     state = "CA",
                     geometry = TRUE) 

ca_pop


pal <- colorQuantile(palette = "viridis", domain = ca_pop$estimate, n = 5)

ca_density <- ca_pop %>%
    st_transform(crs = "+init=epsg:4326") %>%
    leaflet(width = "100%") %>%
    addProviderTiles(provider = "CartoDB.Positron") %>%
    addPolygons(popup = ~ str_extract(NAME, "^([^,]*)"),
                stroke = FALSE,
                smoothFactor = 0,
                fillOpacity = 0.7,
                color = ~ pal(estimate)) %>%
    addLegend("bottomright", 
              pal = pal, 
              values = ~ estimate,
              title = "Population Percentiles",
              opacity = 1)

ca_density



######################################################

#Genomics bit#
#read in the data using bioStrings package
cons_seq <- readDNAStringSet("consensus-seqs.fna")

#number of occurances of each character
alphabetFrequency(cons_seq, baseOnly=TRUE, collapse=TRUE) #total 
#      A       C       G       T   other 
#1847658 1495638 1935723 1460705  342960 


freqs <- alphabetFrequency(DNAStringSet(cons_seq[!is.na(cons_seq)]), baseOnly=TRUE) #by sequence
head(freqs)

writeXStringSet(cons_seq,file="wnv_fasta.fas",format="fasta")


##DECIPHER package for sequence alignment  
library(DECIPHER)
dna <- unique(cons_seq) # the unique input sequences
index <- match(cons_seq, u_dna) # de-replication index
aligned_seq <- AlignSeqs(cons_seq) # align the sequences directly without translation
aligned_seq <- U_DNA[index]
names(cons_seq) <- names(cons_seq) # the re-replicated alignment
head(aligned_seq)

#Post processing 
BrowseSeqs(aligned_seq, highlight=0) #visualize sequences in internet browser

wnv_adjusted <- AdjustAlignment(aligned_seq)  #fixes any "obvious mistakes" 

writeXStringSet(wnv_adjusted,file="wnv_adjusted.fas",format="fasta")

#use shortread to trim off the ends of the data where there are a lot of missing values 
wnv_clean <- DNAStringSet(substr(wnv_adjusted,31,10963))
wnv_clean
BrowseSeqs(wnv_clean, highlight=0)

writeXStringSet(wnv_clean,file="wnv_clean.fas",format="fasta")

#Play with stratag package to filter poor reads using gtypes objects 

library(strataG)
wnv_gtype <- system.file("wnv_clean.fas", package = "strataG")
wnv_gtype <- read.dna(wnv_gtype)



#use pegas to build a haplotype network
input <- "wnv_clean.fas"
fasta.read <- ape::read.dna(input, format='fasta')
distance_wnv <- dist.dna(fasta.read)
wnv_hap <- pegas::haplotype(fasta.read)
wnv_hap <- sort(wnv_hap, what = "label")
(net <- pegas::haploNet(wnv_hap))
ind.hap <-with(stack(setNames(attr(wnv_hap, "index"), rownames(wnv_hap))),
table(hap=ind, pop=rownames(fasta.read)[values])
)
plot(net, size=attr(net, "freq"), scale.ratio=0.2, pie=ind.hap)
legend(-8, 0, colnames(ind.hap), col=rainbow(ncol(ind.hap)), pch=19, ncol=2)
```

********************************************************************************
*Kyle tentatively comparing variation in sequences within a host species*
Potential packages:
Biostrings for alignment and organization
jasonedit-list viewer interactive


