---
title: "â€˜Omics Project"
author: "Aeriel Belk, Amanda Walz, Rebecca Cheek, Kyle Root"
date: "11/7/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This is an R-markdown we can use!!

Steps: 

1. import and early explore data
2. then, see what analyses we can do

```{r}
library(readr)
wn_meta<- read.csv("metadata/metadata.csv")
```
 

Ideas for us to look at. 
Ariel: Build a phylogenetic tree using Bioconductor(phyloseq) to look at divergence and host species
Kyle:Compare genetic viersity between host species and vecotr species
Amanda: sf plot to use lat-long data and group smilar sequences by region
Rebecca: Landscape genetics approach to look at how the urban vs. rural affects genetic structure or rate of gene flow (see https://doi.org/10.1111/mec.14783). 


Do some analysis of the metadata:

 - species 
 - latitude/longitude
 - distribution over counties
 - sequence similarities/ density by region
 
```{r libraries, message=FALSE, warning=FALSE}
library(seqinr)
library(purrr)
library(dplyr)
library(readr)
library(phyloseq)

# if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
# BiocManager::install("Biostrings", version = "3.8")
library(Biostrings)

```

``` {r metadata}
metadata_url <- paste0("https://raw.githubusercontent.com/aeriel-belk/NSCI535-westnile-omics/master/metadata/metadata.csv")

metadataframe <- read_csv(metadata_url)

```

 
```{r genomics}


```

steps to read in:
1) list.files with full names option to get names of all files in folder, save in object (char vector)
2) pipe that vector into map function that does read.fasta (play around with just the first file first)
3) result should be a long list

seqinr -> read.fasta

```{r read-in}
#1) list files into vector
files <- list.files(path = 'consensus_sequences', full.names = TRUE)

#2) make map function to read in fasta files

test <- read.fasta("consensus_sequences/W006_2013-07-10_USA_CA_SanDiego_-116.946_32.885.fa", 
                   as.string = TRUE, 
                   forceDNAtolower = FALSE)

fasta <- map(.x = files, 
                .f = read.fasta(as.string = TRUE, forceDNAtolower = FALSE),
                col_types = cols(.default = "c")) %>% 
  bind_rows()

test <- read.fasta("consensus_sequences/W006_2013-07-10_USA_CA_SanDiego_-116.946_32.885.fa", as.string = TRUE, forceDNAtolower = FALSE)

fasta <- map(.x = files, .f = read.fasta, 
             as.string = TRUE, forceDNAtolower = FALSE)

fasta2 <- readDNAStringSet(files, format="fasta", nrec=-1L, skip=0L, seek.first.rec=FALSE, use.names=TRUE)

writeXStringSet(fasta2, 'consensus-seqs.fna')
#writeXStringSet(LargeDNAStringSet, 'fname.fa')
#write.fasta(sequences, names, file.out, open = "w", nbchar = 60, as.string = FALSE)

```

  After getting it important and *roughly* in a usable format, I tried doing some alignment and clustering with the biostrings package
  
```{r biostring}



```




